### ISSUE TITLE
Generate random 300K identity vectors

### ISSUE THREAD

**Issue by kk1153** [HaiyuWu/Vec2Face#11]
Hi, 

Can you please provide the method and code for generating the 300K synthetic identity vectors of people that do not exist to be able to reproduce the results stated in paper? I could not understand it very clearly based on the paper and the codebase.

Thanks

**Comment by HaiyuWu** [comment]
Sure, it is simply sampling vectors from a Gaussian distribution and dropping those with similarity values larger than a threshold. These are the identity vectors used for generation.

You can use any approach to achieve it. Here is the code I used.

```
import numpy as np
import faiss
from tqdm import tqdm
import faiss.contrib.torch_utils


def generate_random_identity_features(num_vectors=2000000,
                                      dim=512,
                                      similarity_threshold=0.4,
                                      batch_size=50000,  # Increased batch size for GPU
                                      initial_multiplier=3):
    """
    Generate well-separated random identity features using GPU acceleration.

    Parameters:
    - num_vectors: Target number of vectors to generate
    - dim: Dimension of feature vectors
    - similarity_threshold: Maximum allowed cosine similarity between vectors
    - batch_size: Number of vectors to process at once
    - initial_multiplier: Generate initial_multiplier * num_vectors candidates

    Returns:
    - original_vectors: Array of original (un-normalized) well-separated features
    """

    def generate_batch(batch_size, dim):
        """Generate a batch of random vectors and their normalized versions"""
        original_vectors = np.random.normal(0, 1, (batch_size, dim)).astype('float32')
        norms = np.linalg.norm(original_vectors, axis=1, keepdims=True)
        normalized_vectors = original_vectors / norms
        return original_vectors, normalized_vectors

    # Initialize GPU resources
    res = faiss.StandardGpuResources()

    # Configure GPU index
    config = faiss.GpuIndexFlatConfig()
    config.device = 0  # Use first GPU
    config.useFloat16 = False  # Use full precision for accuracy

    # Create GPU index
    index = faiss.GpuIndexFlatIP(res, dim, config)

    # Initialize storage for accepted vectors
    accepted_originals = []
    accepted_normalized = []
    total_processed = 0

    print(f"Generating {num_vectors} well-separated vectors using GPU...")

    with tqdm(total=num_vectors) as pbar:
        while len(accepted_originals) < num_vectors:
            # Generate a batch of candidates
            orig_batch, norm_batch = generate_batch(batch_size, dim)

            if index.ntotal == 0:
                # Add first batch directly
                index.add(norm_batch)
                accepted_originals.extend(orig_batch)
                accepted_normalized.extend(norm_batch)
                pbar.update(len(norm_batch))
                continue

            # Search for nearest neighbors in existing set
            D, _ = index.search(norm_batch, k=1)

            # Filter vectors based on similarity threshold
            dissimilar_mask = D.flatten() < similarity_threshold
            dissimilar_orig = orig_batch[dissimilar_mask]
            dissimilar_norm = norm_batch[dissimilar_mask]

            if len(dissimilar_norm) > 0:
                index.add(dissimilar_norm)
                accepted_originals.extend(dissimilar_orig)
                accepted_normalized.extend(dissimilar_norm)
                pbar.update(len(dissimilar_norm))

            total_processed += batch_size

            # Break if we've processed too many vectors without success
            if total_processed > num_vectors * initial_multiplier:
                print(f"\nWarning: Reached maximum attempts. Generated {len(accepted_originals)} vectors.")
                break

    # Convert to numpy arrays and trim to desired size
    accepted_originals = np.array(accepted_originals[:num_vectors], dtype='float32')
    accepted_normalized = np.array(accepted_normalized[:num_vectors], dtype='float32')

    # Final verification using GPU index
    final_index = faiss.GpuIndexFlatIP(res, dim, config)
    final_index.add(accepted_normalized)
    D, _ = final_index.search(accepted_normalized, k=2)
    max_similarity = np.max(D[:, 1])

    print(f"\nFinal vector count: {len(accepted_originals)}")
    print(f"Maximum similarity between any two vectors: {max_similarity:.4f}")
    print(f"Shape of output: {accepted_originals.shape}")

    # Clean up GPU resources
    del index
    del final_index

    return accepted_originals


# Generate the vectors
vectors = generate_random_identity_features(
    num_vectors=10000000,
    dim=512,
    similarity_threshold=0.3,
    batch_size=50000  # Larger batch size for GPU
)

# Save the results
np.save("4M_random_separated_features.npy", vectors)
```

**Comment by kk1153** [comment]
Thanks. Could you also explain, how you use PCA on MS1MV2 identity vectors in generating random sample identities as stated in the section 3.2 of the paper? How much of real data does it employ?

**Comment by HaiyuWu** [comment]
Hi, this is an example code. Using the WebFace4M identity vectors to learn PCA is better for producing a higher FR accuracy.
```
def generate_normalized_vectors(num, dim):
    identity = np.load("WebFace4M/center_features.npy")
    pca_id = PCA(n_components=512)
    pca_id.fit(identity)

    transformed_data = pca_id.transform(identity)
    mean = np.mean(transformed_data, axis=0)
    cov = np.cov(transformed_data, rowvar=False)
    new_samples_pca_space = np.random.multivariate_normal(mean, cov, size=num_vectors)
    new_ids = pca_id.inverse_transform(new_samples_pca_space).astype('float32')
    return new_ids, new_ids / np.linalg.norm(new_ids, axis=1, keepdims=True)
```

**Comment by kk1153** [comment]
Thanks for your reply. Can you tell which approach is used in the paper
random
Identity vectors as illustrated in previous reply or the one with random
vectors in PCA subspace of real data?

Is there any comparison between these two if you could tell? Thanks

**Comment by HaiyuWu** [comment]
Hi, sorry for the late reply. I was involved in too many things recently. The following code is used for sampling identity vectors from the learned PCA space. You can simply use a pretrained FR model to calculate the center_feature (identity feature) for each identity folder, then use this code to learn PCA and do sampling.

In the vec2face paper, PCA is learned from MS1MV2 identity features to avoid the WebFace4M identity leakage issue as the model was trained on WebFace4M.
```
def generate_normalized_vectors(num, dim):
    identity = np.load("WebFace4M/center_features.npy")
    pca_id = PCA(n_components=512)
    pca_id.fit(identity)

    transformed_data = pca_id.transform(identity)
    mean = np.mean(transformed_data, axis=0)
    cov = np.cov(transformed_data, rowvar=False)
    new_samples_pca_space = np.random.multivariate_normal(mean, cov, size=num_vectors)
    new_ids = pca_id.inverse_transform(new_samples_pca_space).astype('float32')
    return new_ids, new_ids / np.linalg.norm(new_ids, axis=1, keepdims=True)
```

As for random vs. PCA, from my experiment, using PCA can result in better performance on FR accuracy.
I hope I answer your questions well.
